pipeline.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Pipeline is an ordered sequences of game stages which are executed to produce the
+// next game state.
+//
+// If a stage produces an error or an ended game state, the pipeline is halted at that stage.
+type Pipeline interface {
+	// Execute runs the pipeline stages and produces a next game state.
+	//
+	// If any stage produces an error or an ended game state, the pipeline
+	// immediately stops at that stage.
+	//
+	// Errors should be checked and the other results ignored if error is non-nil.
+	//
+	// If the pipeline is already in an error state (this can be checked by calling Err()),
+	// this error will be immediately returned and the pipeline will not run.
+	//
+	// After the pipeline runs, the results will be the result of the last stage that was executed.
+	Execute(*BoardState, Settings, []SnakeMove) (bool, *BoardState, error)
+	// Err provides a way to check for errors before/without calling Execute.
+	// Err returns an error if the Pipeline is in an error state.
+	// If this error is not nil, this error will also be returned from Execute, so it is
+	// optional to call Err.
+	// The idea is to reduce error-checking verbosity for the majority of cases where a
+	// Pipeline is immediately executed after construction (i.e. NewPipeline(...).Execute(...)).
+	Err() error
+}
+
+// pipeline is an implementation of Pipeline
+type pipeline struct {
+	// stages is a list of stages that should be executed from slice start to end
+	stages []StageFunc
+	// if the pipeline has an error
+	err error
+}
+
+// NewPipeline constructs an instance of Pipeline using the global registry.
+// It is a convenience wrapper for NewPipelineFromRegistry when you want
+// to use the default, global registry.
+func NewPipeline(stageNames ...string) Pipeline {
+	return NewPipelineFromRegistry(globalRegistry, stageNames...)
+}
+
+// NewPipelineFromRegistry constructs an instance of Pipeline, using the specified registry
+// of pipeline stage functions.
+//
+// The order of execution for the pipeline stages will correspond to the order that
+// the stage names are provided.
+//
+// Example:
+// 	NewPipelineFromRegistry(r, s, "stage1", "stage2")
+// ... will result in stage "stage1" running first, then stage "stage2" running after.
+//
+// An error will be returned if an unregistered stage name is used (a name that is not
+// mapped in the registry).
+func NewPipelineFromRegistry(registry map[string]StageFunc, stageNames ...string) Pipeline {
+	// this can't be useful and probably indicates a problem
+	if len(registry) == 0 {
+		return &pipeline{err: errors.New("empty registry")}
+	}
+
+	// this also can't be useful and probably indicates a problem
+	if len(stageNames) == 0 {
+		return &pipeline{err: errors.New("no stages")}
+	}
+
+	p := pipeline{}
+	for _, s := range stageNames {
+		fn, ok := registry[s]
+		if !ok {
+			return pipeline{err: errors.New("stage not found")}
```

Comment:
Realized just now that engine could definitely do this internally, since it knows when it's getting the error via rules, and the assumption is that any rules error on game creation is an invalid request from the client.
It's probably more consistent if rules returns a specific error type though?

Reply:
An error type is a good idea. I had originally setup the error message to include the stage name, but then I removed it because I wanted the error to be static for simpler comparison. But it did nag at me that it would be useful to know which stage caused the error. I think an error type would allow the best of both worlds - a convenient way to check for the error type and a clear message.